

```
# This is formatted as code
```

# DMA Fall 22

**Note** : This entire lab will be manually evaluated.

Name : ''
Collaborator : ''

# Lab 4: Neural Networks #


In [ ]:
import pandas as pd
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC

from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.feature_extraction import DictVectorizer

from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV, ParameterGrid

import numpy as np

import warnings
warnings.filterwarnings("ignore")

In [ ]:
!wget http://askoski.berkeley.edu/~zp/lab_4_training.csv
!wget http://askoski.berkeley.edu/~zp/lab_4_test.csv

df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')
df_train.head()

--2022-09-28 12:25:43--  http://askoski.berkeley.edu/~zp/lab_4_training.csv
Resolving askoski.berkeley.edu (askoski.berkeley.edu)... 169.229.192.179
Connecting to askoski.berkeley.edu (askoski.berkeley.edu)|169.229.192.179|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 79177 (77K) [text/csv]
Saving to: ‘lab_4_training.csv.4’

lab_4_training.csv. 100%[===================>]  77.32K  --.-KB/s    in 0.04s   

2022-09-28 12:25:43 (1.90 MB/s) - ‘lab_4_training.csv.4’ saved [79177/79177]

--2022-09-28 12:25:43--  http://askoski.berkeley.edu/~zp/lab_4_test.csv
Resolving askoski.berkeley.edu (askoski.berkeley.edu)... 169.229.192.179
Connecting to askoski.berkeley.edu (askoski.berkeley.edu)|169.229.192.179|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 26519 (26K) [text/csv]
Saving to: ‘lab_4_test.csv.4’

lab_4_test.csv.4    100%[===================>]  25.90K  --.-KB/s    in 0.02s   

2022-09-28 12:25:43 (1.34 MB/s) - ‘lab_4_test.csv.4’ saved [

Unnamed: 0  gender  age    year eyecolor  height  miles  brothers  sisters  \
0         577    male   20   third    hazel    72.0  180.0         0        0   
1         677    male   19  second    hazel    72.0  120.0         1        1   
2        1738    male   20  second    brown    63.0   55.0         1        2   
3        1355    male   20   third    green    78.0  200.0         0        0   
4         891  female   19  second    green    67.0  280.0         2        0   

   computertime exercise  exercisehours  musiccds  playgames  watchtv  
0           5.0       No            0.0     100.0       10.0     10.0  
1          16.0      Yes            9.0      70.0        3.0      5.0  
2          15.0      Yes            4.5      15.0        4.0     13.0  
3          10.0      Yes            9.0      20.0       10.0     10.0  
4           4.0      Yes            2.0     164.0        0.0      2.0

In [ ]:
df_test.head()

Unnamed: 0 gender  age    year eyecolor  height  miles  brothers  sisters  \
0        1303   male   20  second    green    73.0  210.0         0        1   
1          36   male   20   third    other    71.0   90.0         1        0   
2         489   male   22  fourth    hazel    75.0  200.0         0        1   
3        1415   male   19  second    brown    72.0   35.0         2        2   
4         616   male   22  fourth    hazel    71.0   15.0         2        1   

   computertime exercise  exercisehours  musiccds  playgames  watchtv  
0          10.0      Yes            5.0      50.0        1.0     15.0  
1          15.0      Yes            4.0      10.0        0.0      1.0  
2           1.0      Yes            2.0     150.0        1.0     10.0  
3          20.0      Yes            5.0     100.0        0.0      7.0  
4          10.0      Yes            7.0      10.0        0.0      5.0

***
### Question 1###
Calculate a baseline accuracy measure using the majority class, assuming a target variable of `gender`. The majority class is the most common value of the target variable in a particular dataset. Accuracy is calculated as (true positives + true negatives) / (all negatives and positives).

**Question 1.a**  
Find the majority class in the training set. If you always predicted this class in the training set, what would your accuracy be?

In [ ]:
df_train.groupby('gender').count()
true_pos = sum(df_train['gender'] == 'female')
true_neg = 0
total = df_train['gender'].count()
accuracy = (true_pos + true_neg) / total
accuracy

0.5427852348993288

###ANSWER: 0.5427852348993288

**Question 1.b**   
If you always predicted this same class (majority from the training set) in the test set, what would your accuracy be?

In [ ]:
df_test.groupby('gender').count()
true_pos = sum(df_test['gender'] == 'female')
true_neg = 0
total = df_test['gender'].count()
accuracy = (true_pos + true_neg) / total
accuracy

0.5226130653266332

###ANSWER: 0.5226130653266332

***
### Question 2 ###
Get started with Neural Networks.

   
Choose a NN implementation (eg: scikit-learn) and specify which you choose. Be sure the implementation allows you to modify the number of hidden layers and hidden nodes per layer.  

NOTE: When possible, specify the logsig (`sigmoid`/`logistic`) function as the transfer function (another word for activation function) and use Levenberg-Marquardt backpropagation (`lbfgs`). It is possible to specify logistic in Sklearn MLPclassifier (Neural net).  

**Question 2.a**   
Train a neural network with a single 10 node hidden layer. Only use the `height` feature of the dataset to predict the `gender`. You will have to change `gender` to a 0 and 1 class. After training, use your trained model to predict the class (`gender`) using the `height` feature from the training set. What is the accuracy of this prediction?

In [ ]:
df_train_encoded = df_train.replace('female', 1).replace('male', 0)

X_train = df_train_encoded[['height']]
Y_train = df_train_encoded['gender']
clf = MLPClassifier(hidden_layer_sizes=10, activation= 'logistic', max_iter=200, alpha=0.0001, solver='lbfgs', verbose=8, random_state=42)
clf.fit(X_train, Y_train)
y_pred_train = clf.predict(X_train)
a = accuracy_score(Y_train, y_pred_train)
a

0.8439597315436241

###ANSWER: 0.8439597315436241

**Question 2.b**  
Take the trained model from question 2.a and use it to predict the test set. This can be accomplished by taking the trained model and giving it the `height` feature values from the test set. What is the accuracy of this model on the test set?

In [ ]:
df_test_encoded = df_test.replace('female', 1).replace('male', 0)
X_test = df_test_encoded[['height']]
Y_test = df_test_encoded['gender']
y_pred_test = clf.predict(X_test)
b = accuracy_score(Y_test, y_pred_test)
b

0.8542713567839196



```
# This is formatted as code
```

###ANSWER: 0.8542713567839196

**Question 2.c**   
Neural Networks tend to prefer smaller, normalized feature values. Try taking the log of the `height` feature in both training and testing sets or use a Standard Scalar operation in SKlearn to centre and normalize the data between 0-1 for continuous values. Repeat question 2.a and 2.b with the log version or the normalized and centered version of this feature.

In [ ]:
df_train_encoded = df_train.replace('female', 1).replace('male', 0)

X_train_log = np.log(df_train_encoded[['height']])
Y_train_log = df_train_encoded['gender']
clf = MLPClassifier(hidden_layer_sizes=10, activation= 'logistic', max_iter=200, alpha=0.0001, solver='lbfgs', verbose=8, random_state=42)
clf.fit(X_train_log, Y_train_log)
y_pred_train_log = clf.predict(X_train_log)
c = accuracy_score(Y_train_log, y_pred_train_log)
c

df_test_encoded = df_test.replace('female', 1).replace('male', 0)
X_test = np.log(df_test_encoded[['height']])
Y_test = df_test_encoded['gender']
y_pred_test = clf.predict(X_test)
d = accuracy_score(Y_test, y_pred_test)
d
print(c, d)

0.8338926174496645 0.8291457286432161


###ANSWER: repeat for 2a: 0.8338926174496645
### repeat for 2b: 0.8291457286432161



***

### Question 3###
The rest of features in this dataset except a few are categorical. No ML method accepts categorical features, so transform `year`, `eyecolor`, `exercise` into a set of binary features, one feature per unique original feature value, and mark the binary feature as ‘1’ if the feature value matches the original value and ‘0’ otherwise. Using only these binary variable transformed features, train and predict the class of the test set. What was your accuracy using Neural Network with a single 10 node hidden layer? During training, use a maximum number of iterations of 50.

In [ ]:
df_train.head(10)
df_train1 = df_train_encoded[['year', 'eyecolor', 'exercise']]
one_hot = pd.get_dummies(df_train1[['year', 'eyecolor', 'exercise']])
df_train1 = df_train1.drop(['year', 'eyecolor', 'exercise'], axis=1)
df_train1 = df_train1.join(one_hot)

df_test1 = df_test_encoded[['year', 'eyecolor', 'exercise']]
one_hot = pd.get_dummies(df_test1[['year', 'eyecolor', 'exercise']])
df_test1 = df_test1.drop(['year', 'eyecolor', 'exercise'], axis=1)
df_test1 = df_test1.join(one_hot)

c_X_train = df_train1
c_Y_train = df_train_encoded['gender']
c_X_test = df_test1
c_Y_test = df_test_encoded['gender']

clf = MLPClassifier(hidden_layer_sizes=10, activation= 'logistic', solver= 'lbfgs', max_iter= 50, random_state=42)
clf.fit(c_X_train, c_Y_train)
pred_test = clf.predict(c_X_test)
a = accuracy_score(c_Y_test, pred_test)
a


0.5527638190954773

###ANSWER: 0.5527638190954773

***
### Question 4###
Using a NN, report the accuracy on the test set of a model that trained only on `height` and the `eyecolor` features of instances in the training set.

**Question 4.a**  
What is the accuracy on the test set using the original `height` values (no pre-processing) and `eyecolor` as a one-hot?

In [ ]:
df_train1 = df_train_encoded[['eyecolor', 'height']]
one_hot = pd.get_dummies(df_train1[['eyecolor']])
df_train1 = df_train1.drop(['eyecolor'], axis=1)
df_train1 = df_train1.join(one_hot)
df_train1

df_test1 = df_test_encoded[['eyecolor', 'height']]
one_hot = pd.get_dummies(df_test1[['eyecolor']])
df_test1 = df_test1.drop(['eyecolor'], axis=1)
df_test1 = df_test1.join(one_hot)

c_X_train = df_train1
c_Y_train = df_train_encoded['gender']
c_X_test = df_test1
c_Y_test = df_test_encoded['gender']

clf = MLPClassifier(hidden_layer_sizes=10, activation= 'logistic', solver= 'lbfgs', max_iter= 200, random_state=42)
clf.fit(c_X_train, c_Y_train)
pred_test = clf.predict(c_X_test)
a = accuracy_score(c_Y_test, pred_test)
a

0.8618090452261307

###ANSWER: 0.8618090452261307

**Question 4.b**  
What is the accuracy on the test set using the log of `height` values (applied to both training and testing sets) and `eyecolor` as a one-hot?

In [ ]:
df_train_encoded['log_h'] = np.log(df_train_encoded[['height']])

df_train1 = df_train_encoded[['eyecolor', 'log_h']]
one_hot = pd.get_dummies(df_train1[['eyecolor']])
df_train1 = df_train1.drop(['eyecolor'], axis=1)
df_train1 = df_train1.join(one_hot)
df_train1

df_test_encoded['log_h'] = np.log(df_test_encoded[['height']])

df_test1 = df_test_encoded[['eyecolor', 'log_h']]
one_hot = pd.get_dummies(df_test1[['eyecolor']])
df_test1 = df_test1.drop(['eyecolor'], axis=1)
df_test1 = df_test1.join(one_hot)

c_X_train = df_train1
c_Y_train = df_train_encoded['gender']
c_X_test = df_test1
c_Y_test = df_test_encoded['gender']

clf = MLPClassifier(hidden_layer_sizes=10, activation= 'logistic', solver= 'lbfgs', max_iter= 200, random_state=42)
clf.fit(c_X_train, c_Y_train)
pred_test = clf.predict(c_X_test)
a = accuracy_score(c_Y_test, pred_test)
a

0.8618090452261307

###ANSWER: 0.8618090452261307

**Question 4.c**  
What is the accuracy on the test set using the Z-score of `height` values and `eyecolor` as a one-hot?

Z-score is a normalization function. It is the value of a feature minus the average value for that feature (in the training set), divided by the standard deviation of that feature (in the training set). Remember that, whenever applying a function to a feature in the training set, it also has to be applied to that same feature in the test set.

In [ ]:
df_train_encoded['z_h'] = (df_train_encoded['height']- df_train_encoded['height'].mean())/df_train_encoded['height'].std()

df_train1 = df_train_encoded[['eyecolor', 'z_h']]
one_hot = pd.get_dummies(df_train1[['eyecolor']])
df_train1 = df_train1.drop(['eyecolor'], axis=1)
df_train1 = df_train1.join(one_hot)
df_train1

df_test_encoded['z_h'] = (df_test_encoded['height']- df_test_encoded['height'].mean())/df_test_encoded['height'].std()

df_test1 = df_test_encoded[['eyecolor', 'z_h']]
one_hot = pd.get_dummies(df_test1[['eyecolor']])
df_test1 = df_test1.drop(['eyecolor'], axis=1)
df_test1 = df_test1.join(one_hot)

c_X_train = df_train1
c_Y_train = df_train_encoded['gender']
c_X_test = df_test1
c_Y_test = df_test_encoded['gender']

clf = MLPClassifier(hidden_layer_sizes=10, activation= 'logistic', solver= 'lbfgs', max_iter= 200, random_state=42)
clf.fit(c_X_train, c_Y_train)
pred_test = clf.predict(c_X_test)
a = accuracy_score(c_Y_test, pred_test)
a

0.864321608040201

###ANSWER: 0.864321608040201

***
### Question 5 ###
Repeat question 4 for `exercisehours` & `eyecolor`.

In [ ]:
df_train1 = df_train_encoded[['eyecolor', 'exercisehours']]
one_hot = pd.get_dummies(df_train1[['eyecolor']])
df_train1 = df_train1.drop(['eyecolor'], axis=1)
df_train1 = df_train1.join(one_hot)
df_train1

df_test1 = df_test_encoded[['eyecolor', 'exercisehours']]
one_hot = pd.get_dummies(df_test1[['eyecolor']])
df_test1 = df_test1.drop(['eyecolor'], axis=1)
df_test1 = df_test1.join(one_hot)

c_X_train = df_train1
c_Y_train = df_train_encoded['gender']
c_X_test = df_test1
c_Y_test = df_test_encoded['gender']

clf = MLPClassifier(hidden_layer_sizes=10, activation= 'logistic', solver= 'lbfgs', max_iter= 200, random_state=42)
clf.fit(c_X_train, c_Y_train)
pred_test = clf.predict(c_X_test)
a = accuracy_score(c_Y_test, pred_test)
a

0.5728643216080402

In [ ]:
df_train_encoded= df_train_encoded[df_train_encoded['exercisehours'] != 0]
df_train_encoded['log_h'] = np.log(df_train_encoded['exercisehours'])

df_train1 = df_train_encoded[['eyecolor', 'log_h']]
one_hot = pd.get_dummies(df_train1[['eyecolor']])
df_train1 = df_train1.drop(['eyecolor'], axis=1)
df_train1 = df_train1.join(one_hot)
df_train1

df_test_encoded= df_test_encoded[df_test_encoded['exercisehours'] != 0]
df_test_encoded['log_h'] = np.log(df_test_encoded['exercisehours'])

df_test1 = df_test_encoded[['eyecolor', 'log_h']]
one_hot = pd.get_dummies(df_test1[['eyecolor']])
df_test1 = df_test1.drop(['eyecolor'], axis=1)
df_test1 = df_test1.join(one_hot)

c_X_train = df_train1
c_Y_train = df_train_encoded['gender']
c_X_test = df_test1
c_Y_test = df_test_encoded['gender']

clf = MLPClassifier(hidden_layer_sizes=10, activation= 'logistic', solver= 'lbfgs', max_iter= 200, random_state=42)
clf.fit(c_X_train, c_Y_train)
pred_test = clf.predict(c_X_test)
a = accuracy_score(c_Y_test, pred_test)
a

0.5774058577405857

In [ ]:
df_train_encoded['z_h'] = (df_train_encoded['exercisehours']- df_train_encoded['exercisehours'].mean())/df_train_encoded['exercisehours'].std()

df_train1 = df_train_encoded[['eyecolor', 'z_h']]
one_hot = pd.get_dummies(df_train1[['eyecolor']])
df_train1 = df_train1.drop(['eyecolor'], axis=1)
df_train1 = df_train1.join(one_hot)
df_train1

df_test_encoded['z_h'] = (df_test_encoded['exercisehours']- df_test_encoded['exercisehours'].mean())/df_test_encoded['exercisehours'].std()

df_test1 = df_test_encoded[['eyecolor', 'z_h']]
one_hot = pd.get_dummies(df_test1[['eyecolor']])
df_test1 = df_test1.drop(['eyecolor'], axis=1)
df_test1 = df_test1.join(one_hot)

c_X_train = df_train1
c_Y_train = df_train_encoded['gender']
c_X_test = df_test1
c_Y_test = df_test_encoded['gender']

clf = MLPClassifier(hidden_layer_sizes=10, activation= 'logistic', solver= 'lbfgs', max_iter= 200, random_state=42)
clf.fit(c_X_train, c_Y_train)
pred_test = clf.predict(c_X_test)
a = accuracy_score(c_Y_test, pred_test)
a

0.5564853556485355

###ANSWER: 5a: 0.5728643216080402; 5b: 0.5774058577405857; 5c: 0.5564853556485355

***
### Question 6###
Combine the features from question 3, 4, and 5 (`year`, `eyecolor`, `exercise`, `height`, `exercisehours`). For numeric features use the best normalization method from questions 4 and 5.

**Question 6.a**  
What was the NN accuracy on the test set using the single 10 node hidden layer?

In [ ]:
# YOUR CODE HERE
df_train_encoded= df_train_encoded[df_train_encoded['exercisehours'] != 0]
df_train_encoded['log_h'] = np.log(df_train_encoded['height'])
df_train_encoded['log_eh'] = np.log(df_train_encoded['exercisehours'])

df_train1 = df_train_encoded[['eyecolor', 'log_h', 'log_eh', 'year', 'exercise']]
one_hot = pd.get_dummies(df_train1[['eyecolor', 'year', 'exercise']])
df_train1 = df_train1.drop(['eyecolor', 'year', 'exercise'], axis=1)
df_train1 = df_train1.join(one_hot)
df_train1

df_test_encoded= df_test_encoded[df_test_encoded['exercisehours'] != 0]
df_test_encoded['log_h'] = np.log(df_test_encoded['height'])
df_test_encoded['log_eh'] = np.log(df_test_encoded['exercisehours'])

df_test1 = df_test_encoded[['eyecolor', 'log_h', 'log_eh', 'year', 'exercise']]
one_hot = pd.get_dummies(df_test1[['eyecolor', 'year', 'exercise']])
df_test1 = df_test1.drop(['eyecolor', 'year', 'exercise'], axis=1)
df_test1 = df_test1.join(one_hot)

c_X_train = df_train1
c_Y_train = df_train_encoded['gender']
c_X_test = df_test1
c_Y_test = df_test_encoded['gender']

clf = MLPClassifier(hidden_layer_sizes=10, activation= 'logistic', solver= 'lbfgs', max_iter= 200, random_state=43)
clf.fit(c_X_train, c_Y_train)
pred_train = clf.predict(c_X_train)
a = accuracy_score(c_Y_train, pred_train)


clf.fit(c_X_test, c_Y_test)
pred_test = clf.predict(c_X_test)
b = accuracy_score(c_Y_test, pred_test)

clf.fit(c_X_train, c_Y_train)
pred_test = clf.predict(c_X_test)
c = accuracy_score(c_Y_test, pred_test)

print(a, b, c)



0.836 0.8828451882845189 0.803347280334728


###ANSWER: 0.803347280334728 trained on training data and applied to test data

***
### Question 7- Bonus (10%)###
Can you improve your test set prediction accuracy by 5% or more?  

See how close to that milestone of improvement you can get by modifying the tuning parameters of  Neural Networks (the number of hidden layers, number of hidden nodes in each layer, the learning rate aka mu). A great guide to tuning parameters is explained in this guide: http://www.csie.ntu.edu.tw/~cjlin/papers/guide/guide.pdf.

While the guide is specific to SVM and in particular the C and gamma parameters of the RBF kernel, the method applies to generally to any ML technique with tuning parameters.

Please also write a paragraph in a markdown cell below with an explanation of your approach and evaluation metrics.


In [ ]:
# YOUR CODE HERE

###ANSWER: